pkg/build/admission/secretinjector/admission.go

```diff
@@ -0,0 +1,108 @@
+package secretinjector
+
+import (
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/golang/glog"
+	authclient "github.com/openshift/origin/pkg/auth/client"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	"github.com/openshift/origin/pkg/util/urlpattern"
+
+	"k8s.io/kubernetes/pkg/admission"
+	"k8s.io/kubernetes/pkg/api"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	coreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/unversioned"
+	"k8s.io/kubernetes/pkg/client/restclient"
+)
+
+func init() {
+	admission.RegisterPlugin("openshift.io/BuildConfigSecretInjector", func(c clientset.Interface, config io.Reader) (admission.Interface, error) {
+		return &secretInjector{
+			Handler: admission.NewHandler(admission.Create),
+		}, nil
+	})
+}
+
+type secretInjector struct {
+	*admission.Handler
+	restClientConfig restclient.Config
+}
+
+func (si *secretInjector) Admit(attr admission.Attributes) (err error) {
+	bc, ok := attr.GetObject().(*buildapi.BuildConfig)
+	if !ok {
+		return nil
+	}
+
+	if bc.Spec.Source.SourceSecret != nil || bc.Spec.Source.Git == nil {
+		return nil
+	}
+
+	// TODO: Jenkins bootstrapper verifies attr.GetSubresource() and attr.GetResource() - do I need to?
+
+	// TODO: no idea if the impersonation is necessary or correct?
+	impersonatingConfig := si.restClientConfig
+	impersonatingConfig.WrapTransport = func(rt http.RoundTripper) http.RoundTripper {
+		return authclient.NewImpersonatingRoundTripper(attr.GetUserInfo(), si.restClientConfig.WrapTransport(rt))
+	}
+
+	client, err := coreclient.NewForConfig(&impersonatingConfig)
+	if err != nil {
+		glog.V(2).Infof("secretinjector: could not create client: %v", err)
+		return nil
+	}
+
+	namespace := attr.GetNamespace()
+
+	url, err := url.Parse(bc.Spec.Source.Git.URI)
+	if err != nil {
+		glog.V(2).Infof(`secretinjector: buildconfig "%s/%s": URI %q parse failed: %v`, namespace, bc.GetName(), bc.Spec.Source.Git.URI, err)
+		return nil
+	}
+
+	secrets, err := client.Secrets(namespace).List(api.ListOptions{})
+	if err != nil {
+		glog.V(2).Infof("secretinjector: failed to list Secrets: %v", err)
+		return nil
+	}
+
+	patterns := []*urlpattern.URLPattern{}
+	for _, secret := range secrets.Items {
+		if secret.Type == api.SecretTypeBasicAuth && url.Scheme != "http" && url.Scheme != "https" ||
+			secret.Type == api.SecretTypeSSHAuth && url.Scheme != "ssh" {
```

Comment:
you can still use the secret to supply a gitconfig, so you'd still want it to match.  and strictly speaking someone could (incorrectly) create a basicauth type secret that contains a gitconfig key, and we should probably still match/use that secret if the url is appropriate.